store.go

```diff
@@ -1,51 +1,81 @@
 package vstore
 
-// Store is the main redux-like store. Use it to access the State properties
-// through the ComponentConstructor.
-type Store struct {
-	r    Reducer
-	mws  []Middleware
-	subs map[*SubFunc]bool
+import (
+	"github.com/gopherjs/vecty"
+	"sync"
+)
+
+// store is the main redux-like store. Use it to access the State properties.
+type Store interface {
+	Connect(comp StoreComponent) StoreComponent
+	Dispatch(action interface{})
+	State() interface{}
 }
 
-// SubFunc is a subscription callback.
-type SubFunc func(s *Store)
+// store satisfies Store.
+type store struct {
+	r   Reducer
+	mws []Middleware
+
+	mtx   sync.Mutex
+	comps map[*storeComponent]bool
+}
 
 // Middleware is a simple middleware that gets called before any Reduce calls.
 type Middleware func(action interface{})
 
-// CreateStore returns a store that you can make global dispatches to in order
+// New returns a store that you can make global dispatches to in order
 // to update the passed global state.
-func CreateStore(r Reducer, mws ...Middleware) *Store {
-	s := Store{r: r, mws: mws, subs: map[*SubFunc]bool{}}
-
-	return &s
+func New(r Reducer, mws ...Middleware) Store {
+	return &store{r: r, mws: mws, comps: map[*storeComponent]bool{}}
 }
 
 // Dispatch takes an action and passes it to all middlewares and the Reduce function.
-// It also calls any subscription functions if exist.
-func (s *Store) Dispatch(action interface{}) {
+// NOTE: dispatch is called in a goroutine.
+// NOTE: vecty.Rerender is called in a goroutine for each subscribed component.
+func (s *store) Dispatch(action interface{}) {
+	go s.dispatch(action)
+}
+
+// State returns global state.
+func (s *store) State() interface{} {
+	return s.r
+}
+
+// sub subscribes a component to any store updates.
+func (s *store) sub(comp *storeComponent) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.comps[comp] = true
+}
+
+// unsub unsubscribes a component to any store updates.
+func (s *store) unsub(comp *storeComponent) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	delete(s.comps, comp)
+}
+
+// dispatch is called from Dispatch.
+func (s *store) dispatch(action interface{}) {
 	for _, m := range s.mws {
 		m(action)
 	}
 
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
 	s.r.Reduce(action)
 
-	for sub := range s.subs {
-		(*sub)(s)
+	for comp := range s.comps {
+		comp := comp
+		go s.rerender(comp)
 	}
 }
 
-// Subscribe subscribes a callback to any store updates.
-func (s *Store) Subscribe(callback SubFunc) func() {
-	s.subs[&callback] = true
-
-	return func() {
-		delete(s.subs, &callback)
-	}
-}
-
-// GetState returns global state.
-func (s *Store) GetState() interface{} {
-	return s.r
+// rerender connects the store to the component and calls vecty.Rerender on the component.
+func (s *store) rerender(comp *storeComponent) {
```

Comment:
i don't think we need to make this function a `*store` receiver method. Typically, a pointer method exists to update its state, but in this case it's just doing two externally scoped functions. You can either have this be a regular function, or you can inline it on line 73 like so

```go
    for comp := range s.comps {
        go func(comp *storeComponent) {
            comp.Connect(s)
            vecty.Rerender(comp)
        }(comp)
    }
```

Reply:
It's a trivial unexpected method, inline go functions are okay until someone adds complexity.